File: typescript/src/connections.ts
import type { ResolvedConfig } from './config.js'
import { END_USER_HEADER } from './config.js'
import { requestJSON } from './http.js'
import { InvalidRequestError } from './errors.js'
import type { ConnectLink, Connection } from './types.js'

type ConnectionPayload = {
	provider: string
	registry?: string
	code?: string
	status: Connection['status']
	account_ref?: string
	expires_at?: string
}

type ConnectionsPayload = {
	end_user?: string
	actor?: string
	connections?: ConnectionPayload[]
}

type LinkPayload = {
	connect_url: string
	ticket: string
	provider?: string
	expires_at: string
}

export function connectionsPath(slug: string, endUser?: string): string {
	const query = endUser ? `?end_user=${encodeURIComponent(endUser)}` : ''
	return `/${encodeURIComponent(slug)}/connections${query}`
}

export async function listConnections(
	config: ResolvedConfig,
	slug: string,
	endUser?: string,
	signal?: AbortSignal
): Promise<Connection[]> {
	const { body } = await requestJSON<ConnectionsPayload>(config, 'GET', connectionsPath(slug, endUser), {
		signal,
	})
	return (body?.connections ?? []).map(toConnection)
}

export async function createConnectLink(
	config: ResolvedConfig,
	slug: string,
	endUser: string,
	provider: string | undefined,
	signal?: AbortSignal
): Promise<ConnectLink> {
	const { body } = await requestJSON<LinkPayload>(
		config,
		'POST',
		`/${encodeURIComponent(slug)}/connections/links`,
		{ body: { end_user: endUser, ...(provider ? { provider } : {}) }, headers: { [END_USER_HEADER]: endUser }, signal }
	)
	return {
		connectUrl: body.connect_url,
		ticket: body.ticket,
		provider: body.provider || undefined,
		expiresAt: new Date(body.expires_at),
	}
}

function toConnection(payload: ConnectionPayload): Connection {
	return {
		provider: payload.provider,
		registry: payload.registry || undefined,
		code: payload.code || undefined,
		status: payload.status,
		accountRef: payload.account_ref || undefined,
		expiresAt: payload.expires_at ? new Date(payload.expires_at) : undefined,
	}
}

/**
 * The gateway is the authority on what an end-user id may be; this only
 * catches the mistake worth catching locally, which is the empty one — it
 * would otherwise be sent as a header the gateway reads as "no user named"
 * and answer for the wrong actor.
 */
export function requireEndUser(endUser: string): string {
	const trimmed = endUser?.trim() ?? ''
	if (!trimmed) {
		throw new InvalidRequestError('an end-user id is required to act for a user')
	}
	return trimmed
}

File: python/src/trustgate/connections.py
"""The connections API: which upstream accounts an actor has, and the link to add one."""

from __future__ import annotations

import json
from datetime import datetime
from typing import Any
from urllib.parse import quote

from .config import API_KEY_HEADER, END_USER_HEADER, Config
from .errors import InvalidRequestError
from .transport import Transport, error_for_response
from .types import ConnectLink, Connection


def connections_path(slug: str, end_user: str | None = None) -> str:
    query = f"?end_user={quote(end_user)}" if end_user else ""
    return f"/{quote(slug)}/connections{query}"


def list_connections(
    config: Config, transport: Transport, slug: str, end_user: str | None = None
) -> list[Connection]:
    response = transport.request(
        "GET",
        f"{config.base_url}{connections_path(slug, end_user)}",
        {API_KEY_HEADER: config.api_key, "Accept": "application/json"},
        None,
        config.timeout,
    )
    if response.status >= 400:
        raise error_for_response(response)
    body = response.json() or {}
    return [_to_connection(item) for item in (body.get("connections") or [])]


def create_connect_link(
    config: Config,
    transport: Transport,
    slug: str,
    end_user: str,
    provider: str | None = None,
) -> ConnectLink:
    payload: dict[str, Any] = {"end_user": end_user}
    if provider:
        payload["provider"] = provider
    response = transport.request(
        "POST",
        f"{config.base_url}/{quote(slug)}/connections/links",
        {
            API_KEY_HEADER: config.api_key,
            END_USER_HEADER: end_user,
            "Accept": "application/json",
            "Content-Type": "application/json",
        },
        json.dumps(payload).encode("utf-8"),
        config.timeout,
    )
    if response.status >= 400:
        raise error_for_response(response)
    body = response.json() or {}
    return ConnectLink(
        connect_url=body.get("connect_url", ""),
        ticket=body.get("ticket", ""),
        provider=body.get("provider") or None,
        expires_at=_to_datetime(body.get("expires_at")) or datetime.now(),
    )


def require_end_user(end_user: str) -> str:
    """The gateway is the authority on what an end-user id may be.

    This only catches the mistake worth catching locally, which is the empty
    one - it would otherwise be sent as a header the gateway reads as "no user
    named" and answer for the wrong actor.
    """
    trimmed = (end_user or "").strip()
    if not trimmed:
        raise InvalidRequestError("an end-user id is required to act for a user")
    return trimmed


def _to_connection(payload: dict[str, Any]) -> Connection:
    return Connection(
        provider=payload.get("provider", ""),
        status=payload.get("status", ""),
        registry=payload.get("registry") or None,
        code=payload.get("code") or None,
        account_ref=payload.get("account_ref") or None,
        expires_at=_to_datetime(payload.get("expires_at")),
    )


def _to_datetime(raw: Any) -> datetime | None:
    if not raw or not isinstance(raw, str):
        return None
    try:
        return datetime.fromisoformat(raw.replace("Z", "+00:00"))
    except ValueError:
        return None
